lib/services/firewall.pm

```diff
@@ -0,0 +1,75 @@
+# SUSE's openQA tests
+#
+# Copyright © 2021 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: Package for firewall service tests
+#
+# Maintainer: Huajian Luo <hluo@suse.com>
+
+package services::firewall;
+use base 'opensusebasetest';
+use testapi;
+use utils;
+use strict;
+use warnings;
+
+my $service      = 'firewalld';
+my $service_type = 'Systemd';
+my $pkg          = 'SuSEfirewall2';
+
+sub install_service {
+    zypper_call('in ' . $pkg);
+}
+
+sub susefirewall2_to_firewalld {
+    my $timeout = 180;
+    script_run('susefirewall2-to-firewalld -c',                                     timeout => $timeout);
+    script_run('firewall-cmd --permanent --zone=external --add-service=vnc-server', timeout => $timeout);
+    systemctl 'restart firewalld';
+    script_run('iptables -S', timeout => $timeout);
+}
+
+sub enable_service {
+    common_service_action $service, $service_type, 'enable';
+}
+
+sub start_service {
+    common_service_action $service, $service_type, 'start';
+}
+
+# check service is running and enabled
+sub check_service {
+    common_service_action $service, $service_type, 'is-enabled';
+    common_service_action $service, $service_type, 'is-active';
+}
+
+# check firewall service before and after migration
+# stage is :
+#           'before' for SuSEfirewall2 or
+#           'after' for firewalld after system migration.
+sub full_firewall_check {
+    my ($stage) = @_;
+
+    # we just support SLE12 to SLES15 SuSEfirewall2 to firewalld check
+    return if (get_var('ORIGIN_SYSTEM_VERSION') eq '11-SP4');
```

Comment:
Huajian,
Did you add SLES15 into your consideration?

Reply:
Now we just take care of sles12